app/models/user.rb

```diff
@@ -1,2 +1,97 @@
 class User < ApplicationRecord
+
+  has_many :reviews, dependent: :destroy
+  has_many :orders, dependent: :destroy
+  has_one  :image, as: :imageable, dependent: :destroy
+  has_many :addresses, as: :addressable, dependent: :destroy
+
+  validates_uniqueness_of :email, unless: :is_guest?
+
+  devise :database_authenticatable,
+         :registerable,
+         :recoverable,
+         :rememberable,
+         :trackable,
+         :validatable,
+         :confirmable,
+         :omniauthable, :omniauth_providers => [:facebook]
+
+  scope  :without_provider, -> { where(provider: nil, uid: nil) }
+
+  def billing_address
+    addresses.find_by(kind: :billing)
+  end
+
+  def shipping_address
+    addresses.find_by(kind: :shipping)
+  end
+
+  def self.from_omniauth(auth)
+    @auth = auth
+    case
+      when where(email: @auth.info.email).without_provider.present?
+        @user = where(email: @auth.info.email).without_provider.first.update_attributes(
```

Comment:
1.  where(email: @auth.info.email).without_provider used twice. You can move it to scope